File: tests/fixtures/vulnerable/stretch.go
package main

import (
	"crypto/md5"
	"crypto/tls"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"regexp"
)

// Vulnerable: Hardcoded Credentials (CG-020)
const apiKey = "sk-live-9f8e7d6c5b4a3210"

func connect() string {
	password := "SuperSecret123!"
	return password
}

// Vulnerable: Path Traversal (CG-030) — concatenated path
func readUserFile(name string) ([]byte, error) {
	return os.ReadFile("/data/uploads/" + name)
}

// Vulnerable: Path Traversal (CG-030) — Sprintf-built path
func openLog(day string) (*os.File, error) {
	return os.Open(fmt.Sprintf("/var/log/app/%s.log", day))
}

// Vulnerable: SSRF (CG-060) — concatenated URL
func fetchAvatar(userHost string) (*http.Response, error) {
	return http.Get("http://" + userHost + "/avatar.png")
}

// Vulnerable: SSRF (CG-060) — Sprintf-built URL
func callService(endpoint string) (*http.Response, error) {
	return http.Get(fmt.Sprintf("http://internal-api/%s", endpoint))
}

// Vulnerable: Weak Cryptography (CG-021) — crypto/md5
func hashPassword(password string) [16]byte {
	return md5.Sum([]byte(password))
}

// Vulnerable: Sensitive Data Exposure (CG-040) — password logged
func logLogin(password string) {
	log.Printf("login attempt with password: %s", password)
}

// Vulnerable: Security Misconfiguration (CG-050) — TLS verification disabled
func insecureClient() *tls.Config {
	return &tls.Config{InsecureSkipVerify: true}
}

// Vulnerable: Insecure Randomness (CG-022) — math/rand used for a session ID
func generateSessionID() int {
	return rand.Intn(1000000)
}

// Vulnerable: Insecure Regular Expression / ReDoS (CG-023)
var emailPattern = regexp.MustCompile("^([a-zA-Z0-9]+)+@")

// Vulnerable: Open Redirect (CG-025) — redirect target from query param
func goNext(w http.ResponseWriter, r *http.Request) {
	http.Redirect(w, r, r.URL.Query().Get("next"), http.StatusFound)
}

File: tests/fixtures/vulnerable/Stretch.java
package demo;

import java.io.File;
import java.io.ObjectInputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.Random;
import java.util.regex.Pattern;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.xml.parsers.DocumentBuilderFactory;

public class Stretch {

    // Vulnerable: Hardcoded Credentials (CG-020) — field literal
    private String password = "SuperSecret123!";

    // Vulnerable: Hardcoded Credentials (CG-020) — local literal
    public String connect() {
        String apiKey = "sk-live-9f8e7d6c5b4a3210";
        return apiKey;
    }

    // Vulnerable: Path Traversal (CG-030) — concatenated path
    public File readUserFile(String name) {
        return new File("/data/uploads/" + name);
    }

    // Vulnerable: Path Traversal (CG-030) — String.format-built path
    public String openLog(String day) throws Exception {
        return Files.readString(Paths.get(String.format("/var/log/app/%s.log", day)));
    }

    // Vulnerable: SSRF (CG-060) — concatenated URL
    public URL fetchAvatar(String userHost) throws Exception {
        return new URL("http://" + userHost + "/avatar.png");
    }

    // Vulnerable: Weak Cryptography (CG-021)
    public MessageDigest hashPassword() throws Exception {
        return MessageDigest.getInstance("MD5");
    }

    // Vulnerable: Sensitive Data Exposure (CG-040) — password logged
    public void logLogin(String password) {
        logger.info("login attempt with password=" + password);
    }

    // Vulnerable: Insecure Deserialization (CG-041)
    public Object readUntrusted(ObjectInputStream ois) throws Exception {
        return ois.readObject();
    }

    // Vulnerable: Security Misconfiguration (CG-050) — CSRF disabled
    public void configure(HttpSecurity http) throws Exception {
        http.csrf().disable();
    }

    // Vulnerable: Arbitrary File Read/Write (CG-031) — path from request parameter
    public File readRequestedFile(HttpServletRequest request) {
        return new File(request.getParameter("path"));
    }

    // Vulnerable: Cross-Site Scripting (CG-010) — unescaped reflection to response
    public void echoName(HttpServletRequest request, HttpServletResponse response) throws Exception {
        response.getWriter().write(request.getParameter("name"));
    }

    // Vulnerable: Insecure Randomness (CG-022) — java.util.Random for a reset token
    public String generatePasswordResetToken() {
        Random random = new Random();
        return String.valueOf(random.nextLong());
    }

    // Vulnerable: Insecure Regular Expression / ReDoS (CG-023)
    public Pattern emailPattern() {
        return Pattern.compile("^([a-zA-Z0-9]+)+@");
    }

    // Vulnerable: Open Redirect (CG-025) — redirect target from request parameter
    public void goNext(HttpServletRequest request, HttpServletResponse response) throws Exception {
        response.sendRedirect(request.getParameter("next"));
    }

    // Vulnerable: XML External Entity (CG-070) — external DTD loading enabled
    public void parseXml(DocumentBuilderFactory dbf) throws Exception {
        dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", true);
    }
}
